apps/hash-api/src/graphql/resolvers/flows/runs.ts

```diff
@@ -0,0 +1,322 @@
+  }
+
+  const { activityId, activityType, scheduledAt, inputs } =
+    getActivityScheduledDetails(scheduledEvent);
+
+  const startedEvent = events.findLast(
+    (item) =>
+      item.eventType ===
+        temporal.api.enums.v1.EventType.EVENT_TYPE_ACTIVITY_TASK_STARTED &&
+      item.activityTaskStartedEventAttributes?.scheduledEventId?.toString() ===
+        scheduledEventId?.toString(),
+  );
+
+  return {
+    activityId,
+    activityType,
+    scheduledAt,
+    inputs,
+    attempt: startedEvent?.activityTaskStartedEventAttributes?.attempt ?? 1,
+    startedAt: eventTimeIsoStringFromEvent(startedEvent),
+  };
+};
+
+const mapTemporalWorkflowToFlowStatus = async (
+  workflow: WorkflowExecutionInfo,
+  temporalClient: TemporalClient,
+): Promise<FlowRun> => {
+  const handle = temporalClient.workflow.getHandle(
+    workflow.workflowId,
+    workflow.runId,
+  );
+
+  const { events } = await handle.fetchHistory();
+
+  const workflowInputs = parseHistoryItemPayload(
+    events?.find(
+      (event) =>
+        event.eventType ===
+        temporal.api.enums.v1.EventType.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED,
+    )?.workflowExecutionStartedEventAttributes?.input,
+  );
+
+  const workflowOutputs = parseHistoryItemPayload(
+    events?.find(
+      (event) =>
+        event.eventType ===
+        temporal.api.enums.v1.EventType.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED,
+    )?.workflowExecutionCompletedEventAttributes?.result,
+  );
+
+  const stepMap: { [activityId: string]: StepRun } = {};
+
+  if (events?.length) {
+    /*
+     * Walk backwards from the most recent event until we have populated the latest state data for each step
+     */
+    for (let i = events.length - 1; i >= 0; i--) {
+      const event = events[i];
+      if (!event) {
+        throw new Error("Somehow out of bounds for events array");
+      }
+
+      const nonScheduledAttributes =
+        event.activityTaskStartedEventAttributes ||
+        event.activityTaskCompletedEventAttributes ||
+        event.activityTaskCanceledEventAttributes ||
+        event.activityTaskCancelRequestedEventAttributes ||
+        event.activityTaskFailedEventAttributes ||
+        event.activityTaskTimedOutEventAttributes;
+
+      if (
+        !nonScheduledAttributes &&
+        !event.activityTaskScheduledEventAttributes
+      ) {
+        // This is not an activity-related event
+        continue;
+      }
+
+      const {
+        activityId,
+        activityType,
+        attempt,
+        inputs,
+        startedAt,
+        scheduledAt,
+      } = event.activityTaskScheduledEventAttributes
+        ? getActivityScheduledDetails(event)
+        : getActivityStartedDetails(events, nonScheduledAttributes!);
+
+      if (stepMap[activityId]) {
+        // We've already encountered and therefore populated all the details for this step
+        continue;
+      }
+
+      const activityRecord: StepRun = {
+        stepId: activityId,
```

Comment:
Right now it's still a bit unclear to me how we the `activityId` will actually correspond to the `stepId`, from local testing it seems that right now `activityId` is just an incremented integer (probably the order in which activities are run?). It might be possible to customise this if we find a way to use `scheduleActivity`, [which seems to allow setting an `activityId`](https://typescript.temporal.io/api/classes/proto.coresdk.workflow_commands.ScheduleActivity).

Also as you point out in the GQL comment, we may want to customise what this ID looks like for a single definition of a step which is run multiple times for multiple inputs.

This is fine for now, just also saying I expect this definition is maybe something we will want to modify.

Reply:
We can set `activityId` as part of `ActivityOptions`, which as you say you can pass to `scheduleActivity`. It's also the interface passed to `proxyActivities`, which then passes it on to `scheduleActivity` when the activity is called.

Possibly the easiest thing to do is to write our own `proxyActivities` which allows passing additional options for each activity call... would just be a case of exposing `options` and then merging the default options in with the options provided in each call.

[Here's the source](https://github.com/temporalio/sdk-typescript/blob/e142407b08ef14af551f611c22f42f6565124598/packages/workflow/src/workflow.ts#L526) for `proxyActivities`.